### PR TITLE
Add gated photometric L-R consistency loss and KITTI stereo recipe; delete inert StereoScale

### PR DESCRIPTION
**Pull request by Bovey0809**
## Motivation

The s3d disparity branch learns mostly monocular shortcuts: every supervision path a monocular predictor can satisfy gets satisfied monocularly. Dense photometric left-right consistency is the one supervision form that is dense (every P3 cell) and can only be satisfied by true correspondence. On KITTI it reliably grows the live stereo-matching component of the disparity estimate, and — paired with finer cost-volume disparity sampling — converts that into the first seed-confirmed AP gain on this branch.

## What

- **`photometric_lr_loss` / `photometric_disp_loss` (+ `_ssim`)** in `s3d/loss.py`: monodepth-style 0.85·SSIM + 0.15·L1 photometric error against the left view after warping the right view by the predicted dense disparity, computed at FULL image resolution (low-res warping is too blurry to supervise px-level matching), plus edge-aware disparity smoothness. Gated behind `training.photometric_loss` (default **off**), weighted by `training.loss_weights.photometric`, logged as an 8th `photo` loss item.
- **`yolo26-s3d-kitti.yaml`** — opt-in recipe variant: cost-volume `num_bins` 24→48 (~4 px/bin at full res) + `photometric_loss: true` + `lr_distance` weight 2.0→0.2. Train with `epochs=400`.
- **Deleted: `StereoScale` augmentation** (class in `augment.py`, pipeline entry + import in `dataset.py`, ~52 lines): geometrically inert — `StereoLetterBox` exactly cancels its calibration scaling (net calib = imgsz/w for any scale s), so it only ever added interpolation blur. No behavior change.
- Tests: `tests/test_s3d_photo.py` (true disparity beats wrong disparities on a synthetic stereo pair; finite loss+grad with out-of-bounds warps; gating on/off through the `training:` block; recipe yaml builds). Docs reference regenerated.

Reused/consolidated: the existing gated `training:`-block mechanism and `loss_weights` plumbing (no new config surface); one shared photometric core for any disparity-map source.

## Validation (KITTI Car R40 AP3D@0.7 Moderate, from-scratch MuSGD rect [384,1248], 3 seeds per arm at 400 epochs)

| config | Mod@0.7 (3 seeds) |
|---|---|
| default `yolo26-s3d.yaml` | 24.4 / 24.1 / 24.5 (24.3±0.2) |
| photometric alone (lr 0.2, w 4.0) | 22.9 / 22.4 / 21.7 (22.3±0.6) |
| **`yolo26-s3d-kitti.yaml` (photometric + 48 bins)** | **25.5 / 25.9 / 25.8 (25.7±0.2)** |

Neither lever helps alone (finer bins previously regressed KITTI without photometric; photometric alone loses to base at full budget) — the pair is the unit. Counterfactual right-image-shift probes confirm the mechanism: the disparity branch's stereo response rises from ~10% to ~30% of the geometric ground truth.

The loss stays default-off because it requires well-textured imagery: on largely textureless scenes (verified on a constant-dims close-range client dataset, 6/6 runs) the photometric gradient is ambiguous and the weakly-anchored disparity map drifts; the docstring documents this boundary.

Deleted: `StereoScale` class + pipeline entry + import (see above). Net diff is +code overall — justification per repo rule 2: this is a new gated feature (a genuinely new loss function and recipe config); nothing further existed to relocate or replace, and the one piece of s3d augmentation code shown to be dead weight is removed here.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 Adds an opt-in photometric consistency training recipe for YOLO26 Stereo 3D, alongside broad code-style, typing, documentation, and test improvements.

### 📊 Key Changes
- 🆕 Introduces `yolo26-s3d-kitti.yaml`, an opt-in KITTI-focused Stereo 3D configuration with:
  - Finer stereo cost-volume disparity sampling.
  - Dense left-right photometric consistency loss using SSIM, L1, and smoothness terms.
  - Adjusted Stereo 3D loss weights.
- 🧪 Adds comprehensive tests for photometric loss behavior, numerical stability, configuration gating, and KITTI recipe construction.
- 📈 Reports improved KITTI Car 3D detection performance for the new recipe, while retaining the default `yolo26-s3d.yaml` for low-texture datasets.
- 🧹 Applies widespread code-quality cleanup:
  - Replaces `dict(...)` with standard dictionary literals.
  - Simplifies loops, conditionals, imports, exception re-raising, and placeholder method bodies.
  - Improves typing, including `Self` support and safer optional defaults.
- 📚 Updates documentation and examples across SAM, YOLOE, YOLO-World, tracking, DALI, integrations, and custom training.
- 🛠️ Improves data handling, export utilities, callback logic, semantic dataset selection, and KITTI conversion defaults.
- 🗑️ Removes the unused `StereoScale` augmentation and updates S3D API reference pages.

### 🎯 Purpose & Impact
- 🎯 Enables stronger Stereo 3D depth learning on well-textured stereo imagery through dense visual correspondence supervision.
- 🚗 The new KITTI recipe may significantly improve 3D car detection accuracy, but it is intentionally opt-in because photometric supervision can be unstable on textureless data.
- ✅ Existing Stereo 3D workflows remain unchanged when using the default configuration.
- 🧑‍💻 Developers benefit from clearer, more consistent, and more maintainable code with improved error propagation and typing.
- 📖 Users get cleaner examples and more accurate documentation across Ultralytics workflows, with no major API-breaking changes indicated.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Bovey0809, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). This is an automated welcome message, and an engineer will assist with review. Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/yolo/s3d/loss.py [hunk 1]
```diff
@@ -23,6 +23,85 @@ def laplacian_nll(
...
+    disp_full = F.interpolate(disp, size=(H, W), mode="bilinear", align_corners=True)
+    left, right = imgs[:, :3], imgs[:, 3:6]
+
+    ys = torch.linspace(-1, 1, H, device=imgs.device)
+    xs = torch.linspace(-1, 1, W, device=imgs.device)
+    gy, gx = torch.meshgrid(ys, xs, indexing="ij")
+    gx = gx.unsqueeze(0) - 2.0 * disp_full[:, 0]  # sample right at u - d (grid units: delta = 2*d_px/W)
+    grid = torch.stack([gx, gy.unsqueeze(0).expand_as(gx)], dim=-1)  # [B, H, W, 2]
+    warped = F.grid_sample(right, grid, align_corners=True, padding_mode="border")
+    valid = ((gx >= -1) & (gx <= 1)).unsqueeze(1).float()  # [B, 1, H, W]
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/s3d/loss.py:64]: ⚠️ **HIGH**: Masking out-of-bounds samples makes an all-invalid disparity field a zero-cost solution: once `gx` is outside `[-1, 1]` everywhere, `photo` becomes zero, and a spatially constant large disparity also has zero smoothness. This directly encourages background predictions to escape the image rather than learn correspondence, especially because the supervised `lr_distance` loss only covers assigned foreground anchors. Keep border/occlusion handling, but add a penalty or constraint for invalid samples so they cannot eliminate the photometric loss.

File: ultralytics/utils/git.py [hunk 2]
```diff
@@ -127,7 +127,7 @@ def branch(self) -> str | None:
         if not self.is_repo or not self.head or not self.head.startswith("ref: "):
             return None
         ref = self.head[5:].strip()
-        return ref[len("refs/heads/") :] if ref.startswith("refs/heads/") else ref
+        return ref.removeprefix("refs/heads/")
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/git.py:130]: ⚠️ **HIGH**: `str.removeprefix()` was added in Python 3.9, but the package explicitly supports Python 3.8. Accessing `GitInfo.branch` from a normal checkout will raise `AttributeError` under Python 3.8. Retain the compatible prefix handling.

**Suggested change:**
```suggestion
        return ref[len("refs/heads/") :] if ref.startswith("refs/heads/") else ref
```

File: ultralytics/cfg/models/26/yolo26-s3d-kitti.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+# plus dense self-supervised photometric left-right consistency with a weakened supervised
+# lr_distance anchor. Together (and only together) these lift KITTI Car AP3D@0.7 Moderate from
+# 20.1 to 25.7 (3 seeds, 400-epoch MuSGD schedule; train with epochs=400).
+# Requires well-textured imagery — keep the default yolo26-s3d.yaml for low-texture or
+# constant-object-size data, where the photometric term is unstable and adds nothing.
+# A standard 3ch backbone runs on L/R images separately (batch trick: [2B,3,H,W]),
+# features are split at the tap point for cost volume, then left-only features
+# continue through the neck/head for 2D detection.
+# Benefits: 100% pretrained weight compatibility, clean left-only 2D features.
+# Usage: model=yolo26s-s3d.yaml
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/models/26/yolo26-s3d-kitti.yaml]: 📝 **LOW**: The usage example points to `yolo26s-s3d.yaml`, which is the regular stereo config and does not enable this recipe's photometric loss or 48-bin cost volume. Users following it will not run the KITTI variant; point it at this `yolo26-s3d-kitti.yaml` recipe instead.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The photometric-loss wiring, loss reporting, tests, and inert augmentation removal are well scoped. However, the invalid-warp masking introduces a degenerate zero-loss disparity solution, and one compatibility regression breaks the supported Python 3.8 floor.

💬 Posted 3 inline comments

**Comment by UltralyticsAssistant** [comment]
🎉 Merged—thank you @Bovey0809 for this outstanding contribution!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR brings together a powerful opt-in photometric consistency recipe for YOLO26 Stereo 3D, stronger KITTI performance, robust testing, and extensive code-quality and documentation improvements. The thoughtful configuration gating preserves existing workflows while giving well-textured stereo datasets a meaningful accuracy boost.

We truly appreciate the care, depth, and breadth of this work! 🚀
